[RDY] main: usage() improvements

Pyrohh opened the pull request:
The diff between the actual usage() output might be hard to parse, so here's the diff (ignoring whitespace with `diff -w`) of `vim -h` output:

``` diff
--- oldout  2015-01-27 17:07:48.933613881 -0500
+++ newout  2015-01-27 17:07:31.913614073 -0500
@@ -1,9 +1,8 @@
-NVIM 0.0.0-alpha+201501262108 (compiled Jan 27 2015 17:07:42)
-
-usage: vim [arguments] [file ..]       edit specified file(s)
-   or: vim [arguments] -               read text from stdin
-   or: vim [arguments] -t tag          edit file where tag is defined
-   or: vim [arguments] -q [errorfile]  edit file with first error
+Usage:
+  nvim [arguments] [file ...]      Edit specified file(s)
+  nvim [arguments] -               Read text from stdin
+  nvim [arguments] -t <tag>        Edit file where tag is defined
+  nvim [arguments] -q [errorfile]  Edit file with first error

 Arguments:
    --          Only file names after this
@@ -18,17 +17,17 @@
    -M          Modifications in text not allowed
    -b          Binary mode
    -l          Lisp mode
-   -V[N][fname]        Be verbose [level N] [log messages to fname]
+  -V[N][file]           Be verbose [level N][log messages to file]
    -D          Debugging mode
    -n          No swap file, use memory only
    -r          List swap files and exit
-   -r (with file name) Recover crashed session
+  -r <file>             Recover crashed session
    -L          Same as -r
-   -A          start in Arabic mode
-   -H          Start in Hebrew mode
+  -A                    Start in Arabic mode
    -F          Start in Farsi mode
+  -H                    Start in Hebrew mode
    -T <terminal>   Set terminal type to <terminal>
-   -u <vimrc>      Use <vimrc> instead of any .vimrc
+  -u <nvimrc>           Use <nvimrc> instead of any .nvimrc
    --noplugin      Don't load plugin scripts
    -p[N]       Open N tab pages (default: one for each file)
    -o[N]       Open N windows (default: one for each file)
@@ -42,8 +41,8 @@
    -w <scriptout>  Append all typed commands to file <scriptout>
    -W <scriptout>  Write all typed commands to file <scriptout>
    --startuptime <file>    Write startup timing messages to <file>
-   -i <viminfo>        Use <viminfo> instead of .viminfo
+  -i <nviminfo>         Use <nviminfo> instead of .nviminfo
    --api-info      Dump API metadata serialized to msgpack and exit
-   --embed     Use stdin/stdout as a msgpack-rpc channel. This can be used for embedding Neovim into other programs
-   -h  or  --help  Print Help (this message) and exit
+  --embed               Use stdin/stdout as a msgpack-rpc channel
    --version       Print version information and exit
+  -h | --help           Print this help message and exit
```

Commit:
main.c: usage() improvement

- Factor out main_msg() in favor of mch_msg() and manual indentation, as to
  provide a much closer representation to the actual output of '--help'.
  'gcc -E' reveals that main_msg() only consists of 3 printf calls
  anyways.
- Factor out for loop used for printing top part of usage text; just
  print the text normally.

usage() text:
- Don't print the version; that's what '--version' is for.
- Be consistent about nomenclature, e.g. '<arg>' denotes required
  argument, '-h | --help' denotes '-h' and '--help' are equivalent, etc.
- Change some instances of vim{,rc,info} to nvim

Pyrohh commented:
I realized the diff shows a small issue, where I accidentally removed the `[arguments]` fields. I'll fix that when I'm at a computer.

Pyrohh commented:
Repushed and updated the diff.

elmart commented:
I like the increase in clarity these changes bring. Note, though, that this breaks all existing translations for usage message. I think it's a price we can pay now. But it's something that shouldn't go unnoticed.

Pyrohh commented:
Removed `puts` usage in favor of `mch_msg` (until we fully understand it) and rebased/pushed.

elmart commented:
LGTM.

Pyrohh commented:
Any other feedback, or may I mark this as RDY?

elmart commented:
I suppose you mean feedback from other people. 
For what respects to me, it's ready.

justinmk commented on src/nvim/main.c:
```diff
@@ -2057,93 +2057,66 @@ static void mainerr(int n, const char *str)
-/*
- * Print messages for "vim -h" or "vim --help" and exit.
- */
+/// Prints help message and exits
```
Noting "vim -h" or "vim --help" is useful, especially for grepping, but also for zero-friction comprehension of what a function does.

Pyrohh replied:
Noted, I hadn't considered grepping.

Pyrohh commented:
Rebased with the mentioned change.

> I suppose you mean feedback from other people.

Yes, I wasn't clear.

Pyrohh commented:
Marking as ready.

justinmk commented on src/nvim/main.c:
```diff
@@ -2057,93 +2057,66 @@ static void mainerr(int n, const char *str)
-
-  signal_stop();              /* kill us with CTRL-C here, if you like */
-
-  mch_msg(longVersion);
```
Why not print `longVersion`? It's a pretty common tradition with unix tools.

Pyrohh replied:
I haven't seen it done anywhere else, but regardless I think it's just noise. I don't see how it's in any way helpful given that --version and the intro screen already include it.

On February 2, 2015 1:53:02 AM EST, "Justin M. Keyes" <notifications@github.com> wrote:
>>  static void usage(void)
>>  {
>> -  int i;
>> -  static char *(use[]) =
>> -  {
>> -    N_("[file ..]       edit specified file(s)"),
>> -    N_("-               read text from stdin"),
>> -    N_("-t tag          edit file where tag is defined"),
>> -    N_("-q [errorfile]  edit file with first error")
>> -  };
>> -
>> -  signal_stop();              /* kill us with CTRL-C here, if you
>like */
>> -
>> -  mch_msg(longVersion);
>
>Why not print `longVersion`? It's a pretty common tradition with unix
>tools.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/neovim/neovim/pull/1894/files#r23909221

justinmk commented on src/nvim/main.c:
```diff
@@ -2057,93 +2057,66 @@ static void mainerr(int n, const char *str)
+  mch_msg(_("  -n                    No swap file, use memory only\n"));
+  mch_msg(_("  -r                    List swap files and exit\n"));
+  mch_msg(_("  -r <file>             Recover crashed session\n"));
+  mch_msg(_("  -L                    Same as -r\n"));
```
`-L` appears to be a vi equivalent. I don't know why `-r` exists also, but since `-L` does the same thing, there is zero reason to mention it in the already-very-busy `Usage:` message. Of course, we will still continue to support `-L` and document it in `:help`. 

But for `-h` message, eliminating noise is a high priority.

Pyrohh replied:
> Of course, we will still continue to support -L and document it in :help

IMO, dealing with too many arguments via selectively removing them from the help dialogue while still providing documentation is just avoiding the issue.

Why not just remove the ability to use `-r` without arguments, as `-L` is easier to remember. Even better, just remove lisp mode so that `-l` can be used instead. Just to be clear, lisp mode is described as such:
```
-l              Lisp mode.  Sets the 'lisp' and 'showmatch' options on.
```
`lisp` is already set via `runtime/ftplugin/lisp.vim`, and `showmatch` could just be set by the user, whether it be via a local ftplugin or globally.

I could also get rid of `-v`, and just rename this PR & commits to something like `Misc. interface improvements`, but that's up to you.

justinmk replied:
> IMO, dealing with too many arguments via selectively removing them from the help dialogue while still providing documentation is just avoiding the issue.

Avoiding the issue is exactly what we want right now. Removing a redundant option from the "Usage" output is *zero* cost, plus a small net benefit. 

> Why not just remove the ability ....

Because it could break things and it's a low priority with a non-zero cost :)

Pyrohh replied:
> Avoiding the issue is exactly what we want right now

That seems reasonable, I'll remove it (for now :smile:).

justinmk commented on src/nvim/main.c:
```diff
@@ -2057,93 +2057,66 @@ static void mainerr(int n, const char *str)
+  mch_msg(_("  -O[N]                 Like -o but split vertically\n"));
+  mch_msg(_("  +                     Start at end of file\n"));
+  mch_msg(_("  +<lnum>               Start at line <lnum>\n"));
+  mch_msg(_("  --cmd <command>       Execute <command> before loading any vimrc file\n"));
```
nvimrc

Pyrohh replied:
:+1:

Commit:
main.c: Review suggestions

Pyrohh commented:
Repushed

Pyrohh commented:
Repushed.

justinmk commented:
LGTM after squash

Pyrohh commented:
Sounds good, but just one question: wouldn't it be easier for you if you pulled my branch, squash my commits, and then cherry picked onto master instead? Using `hub` makes this trivial; just `hub checkout [this PRs URL]` and it will add my repo as a remote, fetch it, and checkout into it. It's very useful for reviews as well.

I'm not trying to avoid work, just saying that if I squashed my commits it would create another commit on GH for you to review, presumably one you already have (as something could have been snuck in the commit under the guise of a rebase). If you pulled my branch and squashed yourself, then you wouldn't need to re-review the commits (right..?), although then again this really just offloads the trust from me to you. Additionally, you'd avoid a merge commit for only one commit, which I've noticed you've been doing recently.

I'm not trying to start some big debate, so feel free to not take this to heart. Just wanted to put it out there.

On February 3, 2015 1:23:12 PM EST, "Justin M. Keyes" <notifications@github.com> wrote:
>LGTM after squash
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/neovim/neovim/pull/1894#issuecomment-72705027

justinmk commented:
In this case, sure, I asked out of habit, and it was no problem to squash. But in general, authors should reorder/squash/organize their own commits, otherwise we'll end up with a bottleneck situation same as Vim, where Bram does all that stuff himself.

justinmk commented:
Also, rebasing and repushing is a good habit for us to follow because it can have unexpected consequences that should be re-checked by travis.

Pyrohh commented:
I understand, I wouldn't want such a bottleneck situation.
